src/QsCompiler/QirGeneration/Subtransformations/TypeTransformation.cs

```diff
@@ -65,7 +68,9 @@ internal ITypeRef LlvmTypeFromQsharpType(ResolvedType resolvedType)
 
             public override QsResolvedTypeKind OnArrayType(ResolvedType b)
             {
-                this.BuiltType = this.QirTypes.Array;
+                this.BuiltType = this.namedLlvmArrays is null
+                    ? this.QirTypes.Array
+                    : this.QirTypes.NamedLlvmArray($"ArrayStorage{++this.namedLlvmArrays}", this.LlvmTypeFromQsharpType(b));
```

Comment:
(nit)
For readability I would recommend (if that compiles):
```suggestion
                    : this.QirTypes.NamedLlvmArray($"ArrayStorage{++(this.namedLlvmArrays)}", this.LlvmTypeFromQsharpType(b));
```